tests/transforms/test_hamiltonian_expand.py

```diff
@@ -246,7 +247,252 @@ def test_hamiltonian_dif_tensorflow(self):
                     qml.CNOT(wires=[2, 0])
                 qml.expval(H)
 
-            tapes, fn = qml.transforms.hamiltonian_expand(tape)
+            tapes, fn = hamiltonian_expand(tape)
+            res = fn(qml.execute(tapes, dev, qml.gradients.param_shift, interface="tf"))
+
+            assert np.isclose(res, output)
+
+            g = gtape.gradient(res, var)
+            assert np.allclose(list(g[0]) + list(g[1]), output2)
+
+
+with QuantumTape() as s_tape1:
+    qml.PauliX(0)
+    S1 = qml.s_prod(1.5, qml.prod(qml.PauliZ(0), qml.PauliZ(1)))
+    qml.expval(S1)
+    qml.expval(S1)
+
+with QuantumTape() as s_tape2:
+    qml.Hadamard(0)
+    qml.Hadamard(1)
+    qml.PauliZ(1)
+    qml.PauliX(2)
+    S2 = qml.op_sum(
+        qml.prod(qml.PauliX(0), qml.PauliZ(2)),
+        qml.s_prod(3, qml.PauliZ(2)),
+        qml.s_prod(-2, qml.PauliX(0)),
+        qml.PauliX(2),
+        qml.prod(qml.PauliZ(0), qml.PauliX(1)),
+    )
+    qml.expval(S2)
+    qml.expval(S2)
+
+S3 = qml.op_sum(
+    qml.s_prod(1.5, qml.prod(qml.PauliZ(0), qml.PauliZ(1))), qml.s_prod(0.3, qml.PauliX(1))
+)
+
+with QuantumTape() as s_tape3:
+    qml.PauliX(0)
+    qml.expval(S3)
+    qml.expval(qml.PauliX(1))
+    qml.expval(S3)
+
+
+S4 = (
+    qml.prod(qml.PauliX(0), qml.PauliZ(2))
+    + qml.s_prod(3, qml.PauliZ(2))
+    - qml.s_prod(2, qml.PauliX(0))
+    + qml.PauliZ(2)
+    + qml.PauliZ(2)
+    + qml.prod(qml.PauliZ(0), qml.PauliX(1), qml.PauliY(2))
+)
+
+with QuantumTape() as s_tape4:
+    qml.Hadamard(0)
+    qml.Hadamard(1)
+    qml.PauliZ(1)
+    qml.PauliX(2)
+
+    qml.expval(S4)
+    qml.expval(qml.PauliX(2))
+    qml.expval(S4)
+    qml.expval(qml.PauliX(2))
+
+SUM_TAPES = [s_tape1, s_tape2, s_tape3, s_tape4]
+SUM_OUTPUTS = [[-1.5, -1.5], [-6, -6], [-1.5, 0.0, -1.5], [-8, 0, -8, 0]]
+
+
+class TestSumExpand:
+    """Tests for the sum_expand transform"""
```

Comment:
Can we add a test to see if this works for the new return types? It's as simple as doing
```python
def test(self):
    try:
        qml.enable_return()
        [test goes here]
    finally:
        qml.disable_return()

```

Reply:
Done! I did it and had to do a small change.

The only tests that failed where the ones using tf. Could it be that this case is not yet treated with the new return types? Because the test for the `hamiltonian_expand` function with the tf interface also failed.

Reply:
Can you link the failure for the `hamiltonian_expand` function for the TF interface? If I recall correctly those tests passed in the PR with the TF integration.

Reply:
With the last updates only the test from `hamiltonian_expand` is failing... 

Here is the log:
```
./tests/transforms/test_hamiltonian_expand.py::TestHamiltonianExpand::test_hamiltonian_dif_tensorflow Failed: [undefined]AttributeError: 'numpy.float64' object has no attribute '_id'
self = <test_hamiltonian_expand.TestHamiltonianExpand object at 0x28a02fdc0>

    @pytest.mark.tf
    def test_hamiltonian_dif_tensorflow(self):
        """Tests that the hamiltonian_expand tape transform is differentiable with the Tensorflow interface"""
    
        import tensorflow as tf
    
        H = qml.Hamiltonian(
            [-0.2, 0.5, 1], [qml.PauliX(1), qml.PauliZ(1) @ qml.PauliY(2), qml.PauliZ(0)]
        )
        var = tf.Variable([[0.1, 0.67, 0.3], [0.4, -0.5, 0.7]], dtype=tf.float64)
        output = 0.42294409781940356
        output2 = [
            9.68883500e-02,
            -2.90832724e-01,
            -1.04448033e-01,
            -1.94289029e-09,
            3.50307411e-01,
            -3.41123470e-01,
        ]
    
        with tf.GradientTape() as gtape:
            with QuantumTape() as tape:
                for i in range(2):
                    qml.RX(var[i, 0], wires=0)
                    qml.RX(var[i, 1], wires=1)
                    qml.RX(var[i, 2], wires=2)
                    qml.CNOT(wires=[0, 1])
                    qml.CNOT(wires=[1, 2])
                    qml.CNOT(wires=[2, 0])
                qml.expval(H)
    
            tapes, fn = hamiltonian_expand(tape)
            res = fn(qml.execute(tapes, dev, qml.gradients.param_shift, interface="tf"))
    
            assert np.isclose(res, output)
    
>           g = gtape.gradient(res, var)

tests/transforms/test_hamiltonian_expand.py:263: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
.venv/lib/python3.9/site-packages/tensorflow/python/eager/backprop.py:1100: in gradient
    flat_grad = imperative_grad.imperative_grad(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

tape = <tensorflow.python.eager.tape.Tape object at 0x28a0716d0>
target = [0.42294409781940356]
sources = [<tf.Tensor: shape=(), dtype=resource, value=<ResourceHandle(name="Resource-0-at-0x28968f8f0", device="/job:localhost/.../device:CPU:0", container="Anonymous", type="tensorflow::Var", dtype and shapes : "[ DType enum: 2, Shape: [2,3] ]")>>]
output_gradients = None
sources_raw = [<tf.Variable 'Variable:0' shape=(2, 3) dtype=float64, numpy=
array([[ 0.1 ,  0.67,  0.3 ],
       [ 0.4 , -0.5 ,  0.7 ]])>]
unconnected_gradients = <UnconnectedGradients.NONE: 'none'>

    def imperative_grad(tape,
                        target,
                        sources,
                        output_gradients=None,
                        sources_raw=None,
                        unconnected_gradients=UnconnectedGradients.NONE):
      """Computes gradients from the imperatively defined tape on top of the stack.
    
      Works by filtering the tape, computing how many downstream usages are of each
      tensor and entry, and repeatedly applying backward functions until we have
      gradients for all sources.
    
      Args:
       tape: the gradient tape which stores the trace.
       target: either a Tensor or list of Tensors to be differentiated.
       sources: list of Tensors for which we want gradients
       output_gradients: if not None, a list of gradient provided for each Target,
        or None if we are to use the target's computed downstream gradient.
       sources_raw: if not None, a list of the source python objects from which the
        sources were generated. Should have the same length as sources. Only needs
        to be populated if unconnected_gradients is 'zero'.
       unconnected_gradients: determines the value returned if the target and
        sources are unconnected. When 'none' the value returned is None wheras when
        'zero' a zero tensor in the same shape as the sources is returned.
    
      Returns:
       the gradient wrt each of the sources.
    
      Raises:
        ValueError: if the arguments are invalid.
        RuntimeError: if something goes wrong.
      """
      try:
        unconnected_gradients = UnconnectedGradients(unconnected_gradients)
      except ValueError:
        raise ValueError(
            "Unknown value for unconnected_gradients: %r" % unconnected_gradients)
    
>     return pywrap_tfe.TFE_Py_TapeGradient(
          tape._tape,  # pylint: disable=protected-access
          target,
          sources,
          output_gradients,
          sources_raw,
          compat.as_str(unconnected_gradients.value))
E     AttributeError: 'numpy.float64' object has no attribute '_id'

.venv/lib/python3.9/site-packages/tensorflow/python/eager/imperative_grad.py:67: AttributeError
```

Reply:
That's strange 🤔 . The test passes locally and also passed in the relevant PR with that change. 

Maybe there are some issues with versioning again? Does it pass in the CI checks?

Reply:
Then it might be related to my tensorflow version, because I am using the `tensorflow-macos` package.

Reply:
Was this resolved?

Reply:
Yep, all tests now pass with the new return types.